In [1]:
%matplotlib inline
import pandas as pd

In [2]:
titles = pd.read_csv('C:/Rajshekar/LetsUpgrade/Assignments/Datasets/titles.csv')
cast = pd.read_csv('C:/Rajshekar/LetsUpgrade/Assignments/Datasets/cast.csv')
release_dates= pd.read_csv('C:/Rajshekar/LetsUpgrade/Assignments/Datasets/release_dates.csv')
titles.head()

title  year
0    Tasveer Mere Sanam  1996
1              Only You  1994
2  El pueblo del terror  1970
3               Machine  2007
4                  MARy  2008

### How many movies are listed in the titles dataframe?

In [3]:
titles['title'].value_counts

<bound method IndexOpsMixin.value_counts of 0                          Tasveer Mere Sanam
1                                    Only You
2                        El pueblo del terror
3                                     Machine
4                                        MARy
                         ...                 
244909    Black Butterfly in a Colorful World
244910          Hua fei hua wu chun man cheng
244911                   Nippon dabi katsukyu
244912          Under Siege 2: Dark Territory
244913              She Must Be Seeing Things
Name: title, Length: 244914, dtype: object>

### They are 244914 movies listed in the titles dataframe

### What are the earliest two films listed in the titles dataframe?

In [4]:
titles['year'].value_counts

<bound method IndexOpsMixin.value_counts of 0         1996
1         1994
2         1970
3         2007
4         2008
          ... 
244909    2018
244910    1980
244911    1970
244912    1995
244913    1987
Name: year, Length: 244914, dtype: int64>

In [5]:
titles['year'].describe()

count    244914.000000
mean       1986.616012
std          29.459912
min        1894.000000
25%        1967.000000
50%        1997.000000
75%        2012.000000
max        2115.000000
Name: year, dtype: float64

In [6]:
current_year = 2020
titles[(titles['year']>current_year)]

title  year
3442        Infinite Payback 2: Our Darkest Hour  2022
3755                     The Mahabharata: Part 2  2022
5052                                Pat & Norman  2022
7285                                     Au Pair  2022
7991                          The Cat in the Hat  2021
...                                          ...   ...
226471  Puss in Boots 2: Nine Lives & 40 Thieves  2021
230196                                Salton Sea  2021
236806                               Escort Girl  2021
239815      Thee Recluse: The Lost Art of Murder  2021
242196                                 100 Years  2115

[66 rows x 2 columns]

## Thee Recluse: The Lost Art of Murder and 100 Years are the earliest two films listed in the titles dataframe

### How many movies have the title "Hamlet"?

In [7]:
titles[titles['title']=="Hamlet"].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 20 entries, 6422 to 202025
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   title   20 non-null     object
 1   year    20 non-null     int64 
dtypes: int64(1), object(1)
memory usage: 480.0+ bytes


## They are 20 movies have the title "Hamlet"

### How many movies are titled "North by Northwest"?

In [8]:
titles[titles['title']=="North by Northwest"]

title  year
74398  North by Northwest  1959

## There is only one movie titled North by Northwest

### When was the first movie titled "Hamlet" made?

In [9]:
titles[titles['title']=="Hamlet"].describe()

year
count    20.000000
mean   1974.700000
std      37.601932
min    1910.000000
25%    1952.500000
50%    1981.500000
75%    2009.500000
max    2017.000000

## In 1910 the first movie titled "Hamlet" was made.

### List all of the "Treasure Island" movies from earliest to most recent.

In [10]:
titles[titles['title']=='Treasure Island'].sort_values(['year'])

title  year
55686   Treasure Island  1918
196514  Treasure Island  1920
141822  Treasure Island  1934
77354   Treasure Island  1950
65051   Treasure Island  1972
37238   Treasure Island  1973
173051  Treasure Island  1985
27076   Treasure Island  1999

### How many movies were made in the year 1950?

In [11]:
y = 1950
titles[(titles['year']==y)].describe()

year
count  1113.0
mean   1950.0
std       0.0
min    1950.0
25%    1950.0
50%    1950.0
75%    1950.0
max    1950.0

## There are 1113 movies made in 1950

### How many movies were made in the year 1960?

In [12]:
ye = 1960
titles[(titles['year']==ye)].describe()

year
count  1549.0
mean   1960.0
std       0.0
min    1960.0
25%    1960.0
50%    1960.0
75%    1960.0
max    1960.0

## There are 1549 movies made in 1960

### How many movies were made from 1950 through 1959?

In [13]:
y1 = 1950
y2 = 1959
len(titles[(titles['year']>=y1) & (titles['year']<=y2)])

12934

## There are 12934 movies made between 1950 and 1959

### In what years has a movie titled "Batman" been released?

In [14]:
titles[titles['title']=='Batman'].sort_values(['title'])

title  year
39349   Batman  1943
240896  Batman  1989

## In 1943 and 1989 years the movies "Batman" was released

### How many roles were there in the movie "Inception"?

In [15]:
release_dates.head()

title  year      country        date
0   #73, Shaanthi Nivaasa  2007        India  2007-06-15
1                   #BKKY  2016     Cambodia  2017-10-12
2                 #Beings  2015      Romania  2015-01-29
3               #Captured  2017          USA  2017-09-05
4  #Ewankosau saranghaeyo  2015  Philippines  2015-01-21

In [16]:
cast.head()

title  year      name   type                character     n
0        Closet Monster  2015  Buffy #1  actor                  Buffy 4  31.0
1       Suuri illusioni  1985    Homo $  actor                   Guests  22.0
2   Battle of the Sexes  2017   $hutter  actor          Bobby Riggs Fan  10.0
3  Secret in Their Eyes  2015   $hutter  actor          2002 Dodger Fan   NaN
4            Steve Jobs  2015   $hutter  actor  1988 Opera House Patron   NaN

In [17]:
cast[cast['title']=='Inception'].drop_duplicates(subset=None, keep='first', inplace=False)

title  year                  name     type  \
98153    Inception  2010  Michael (III) August    actor   
149008   Inception  2010          Peter Basham    actor   
183937   Inception  2010          Tom Berenger    actor   
197877   Inception  2010              Ken Bhan    actor   
260128   Inception  2010        Virgile Bramly    actor   
...            ...   ...                   ...      ...   
3387147  Inception  2010            Ellen Page  actress   
3444628  Inception  2010        Nicole Pulliam  actress   
3473041  Inception  2010    Lisa (II) Reynolds  actress   
3480170  Inception  2010         Talulah Riley  actress   
3731263  Inception  2010        Shannon Welles  actress   

                     character     n  
98153            LAX Passenger   NaN  
149008   Fischer's Jet Captain  44.0  
183937                Browning   8.0  
197877        Mombasan Gambler   NaN  
260128          Bridge Sub Con  29.0  
...                        ...   ...  
3387147                Ariadne   3.0  
3444628          Lobby Sub Con  43.0  
3473041          Private Nurse  48.0  
3480170                 Blonde  25.0  
3731263                Old Mal  51.0  

[78 rows x 6 columns]

## 76 roles were there in the movie "Inception"

### How many roles in the movie "Inception" are NOT ranked by an "n" value?

In [18]:
cast[cast['title']=='Inception']['n'].isnull().sum()

27

## They are 27 roles in the movie "Inception" are NOT ranked by an "n" value

### But how many roles in the movie "Inception" did receive an "n" value?

In [19]:
cast[cast['title']=='Inception']['n'].describe()

count    51.000000
mean     26.000000
std      14.866069
min       1.000000
25%      13.500000
50%      26.000000
75%      38.500000
max      51.000000
Name: n, dtype: float64

## 51 roles roles in the movie "Inception" did receive an "n" value

### Display the cast of "North by Northwest" in their correct "n"-value order, ignoring roles that did not earn a numeric "n" value.

In [20]:
cast[cast['title']=='North by Northwest'].sort_values('n')

title  year                 name     type  \
871426   North by Northwest  1959           Cary Grant    actor   
3518333  North by Northwest  1959      Eva Marie Saint  actress   
1457165  North by Northwest  1959          James Mason    actor   
3157232  North by Northwest  1959  Jessie Royce Landis  actress   
356707   North by Northwest  1959       Leo G. Carroll    actor   
...                     ...   ...                  ...      ...   
3441077  North by Northwest  1959      Maudie Prickett  actress   
3577128  North by Northwest  1959          Doris Singh  actress   
3599912  North by Northwest  1959         Helen Spring  actress   
3739133  North by Northwest  1959        Susan Whitney  actress   
3751311  North by Northwest  1959       Paula Winslowe  actress   

                                 character    n  
871426                  Roger O. Thornhill  1.0  
3518333                        Eve Kendall  2.0  
1457165                    Phillip Vandamm  3.0  
3157232                    Clara Thornhill  4.0  
356707                       The Professor  5.0  
...                                    ...  ...  
3441077                       Elsie - Maid  NaN  
3577128  First United Nations Receptionist  NaN  
3599912                             Bidder  NaN  
3739133                          Attendant  NaN  
3751311                   Woman at Auction  NaN  

[116 rows x 6 columns]

### Display the entire cast, in "n"-order, of the 1972 film "Sleuth".

In [21]:
cast[cast['title']=='Sleuth'].sort_values('n')

title  year                 name     type  \
326774   Sleuth  2007        Michael Caine    actor   
1705394  Sleuth  1972     Laurence Olivier    actor   
326773   Sleuth  1972        Michael Caine    actor   
1293634  Sleuth  2007             Jude Law    actor   
373876   Sleuth  1972       Alec Cawthorne    actor   
1805907  Sleuth  2007        Harold Pinter    actor   
1465483  Sleuth  1972   John (II) Matthews    actor   
2724553  Sleuth  1972   Eve (III) Channing  actress   
1448690  Sleuth  1972         Teddy Martin    actor   
260242   Sleuth  2007      Kenneth Branagh    actor   
373877   Sleuth  2007  Alec (II) Cawthorne    actor   
2724552  Sleuth  2007    Eve (II) Channing  actress   
3370424  Sleuth  2007    Carmel O'Sullivan  actress   

                          character    n  
326774                       Andrew  1.0  
1705394                 Andrew Wyke  1.0  
326773                  Milo Tindle  2.0  
1293634                        Milo  2.0  
373876            Inspector Doppler  3.0  
1805907                 Man on T.V.  3.0  
1465483  Detective Sergeant Tarrant  4.0  
2724553             Marguerite Wyke  5.0  
1448690      Police Constable Higgs  6.0  
260242            Other Man on T.V.  NaN  
373877            Inspector Doppler  NaN  
2724552             Marguerite Wyke  NaN  
3370424                      Maggie  NaN

### Now display the entire cast, in "n"-order, of the 2007 version of "Sleuth".

In [22]:
masthmovie = cast[cast['title']=='Sleuth']
masthmovie[masthmovie['year']==2007].sort_values('n')

title  year                 name     type          character    n
326774   Sleuth  2007        Michael Caine    actor             Andrew  1.0
1293634  Sleuth  2007             Jude Law    actor               Milo  2.0
1805907  Sleuth  2007        Harold Pinter    actor        Man on T.V.  3.0
260242   Sleuth  2007      Kenneth Branagh    actor  Other Man on T.V.  NaN
373877   Sleuth  2007  Alec (II) Cawthorne    actor  Inspector Doppler  NaN
2724552  Sleuth  2007    Eve (II) Channing  actress    Marguerite Wyke  NaN
3370424  Sleuth  2007    Carmel O'Sullivan  actress             Maggie  NaN

### How many roles were credited in the silent 1921 version of Hamlet?

In [23]:
ra = cast[cast['title']=='Hamlet']
ra[ra['year']==1921].describe()

year         n
count     9.0  9.000000
mean   1921.0  5.000000
std       0.0  2.738613
min    1921.0  1.000000
25%    1921.0  3.000000
50%    1921.0  5.000000
75%    1921.0  7.000000
max    1921.0  9.000000

## They are 9 roles in 1921 version of Hamlet

### How many roles were credited in Branagh’s 1996 Hamlet?

In [24]:
ra = cast[cast['title']=='Hamlet']
ra[ra['year']==1996].describe()

year          n
count    54.0  46.000000
mean   1996.0  23.739130
std       0.0  13.744388
min    1996.0   1.000000
25%    1996.0  12.250000
50%    1996.0  23.500000
75%    1996.0  34.750000
max    1996.0  47.000000

## 54 roles were credited in Branagh’s 1996 Hamlet

### How many "Hamlet" roles have been listed in all film credits through history?

In [25]:
fd = cast[cast['title']== "Hamlet"].drop_duplicates(subset=None, keep='first', inplace=False)

In [26]:
fd

title  year              name     type              character     n
1731     Hamlet  1996        Riz Abbasi    actor  Attendant to Claudius   1.0
7789     Hamlet  1921  Fritz Achterberg    actor            Fortinbras,   9.0
11728    Hamlet  2009      Hayden Adams    actor                Laertes   7.0
11729    Hamlet  2009      Hayden Adams    actor                 Player   7.0
14493    Hamlet  1913       Eric Adeney    actor               Reynaldo  14.0
...         ...   ...               ...      ...                    ...   ...
3673234  Hamlet  1969    Jennifer Tudor  actress             Court lady  23.0
3697397  Hamlet  2000      Diane Venora  actress               Gertrude   3.0
3728751  Hamlet  1996     Perdita Weeks  actress          Second Player  44.0
3733637  Hamlet  2015   Katie (II) West  actress                Ophelia   7.0
3751232  Hamlet  1996      Kate Winslet  actress                Ophelia  46.0

[342 rows x 6 columns]

## They are 342 roles roles have been listed in all film credits through history

### How many people have played an "Ophelia"?

In [27]:
cast[cast['character']== "Ophelia"]

title  year  \
248251                          The Merry Maids of Madness  2016   
410247                                      Jake Gets Paid  2009   
470859                                          Inoperable  2017   
866334                     Dogg's Hamlet, Cahoot's Macbeth  2005   
1364457                                                 11  2014   
...                                                    ...   ...   
3722298                                           D.C. Cab  1983   
3723039                                           Creature  2011   
3733637                                             Hamlet  2015   
3748118  National Theatre Live: Rosencrantz & Guildenst...  2017   
3751232                                             Hamlet  1996   

                         name     type character     n  
248251            Amy Bourque    actor   Ophelia  11.0  
410247         Josie Chisholm    actor   Ophelia   NaN  
470859        Crystal Cordero    actor   Ophelia   NaN  
866334          Lucas Grabeel    actor   Ophelia   NaN  
1364457   Catherine (IV) Lord    actor   Ophelia   NaN  
...                       ...      ...       ...   ...  
3722298       Marsha Warfield  actress   Ophelia   7.0  
3723039  Jennifer Lynn Warren  actress   Ophelia  12.0  
3733637       Katie (II) West  actress   Ophelia   7.0  
3748118     Helen (IV) Wilson  actress   Ophelia   NaN  
3751232          Kate Winslet  actress   Ophelia  46.0  

[117 rows x 6 columns]

## They are 117 people have played an "Ophelia"

### How many people have played a role called "The Dude"?

In [28]:
cast[cast['character']== "The Dude"].describe()

year          n
count    19.000000  12.000000
mean   1992.736842  14.416667
std      28.656669  10.874978
min    1926.000000   1.000000
25%    1986.500000   8.750000
50%    2004.000000  10.500000
75%    2012.500000  16.500000
max    2016.000000  37.000000

## They are 19 people have played a role called "The Dude"

### How many people have played a role called "The Stranger"?

In [29]:
cast[cast['character']== "The Stranger"]

title  year                       name     type  \
25954                       Theeb  2014  Hassan Mutlag Al-Maraiyeh    actor   
35748              My Time to Die  2014       Vincent H. Alexander    actor   
46418                      Hickok  2017               Terral Altom    actor   
67421                    Get Mean  1975               Tony Anthony    actor   
67424    Lo straniero di silenzio  1968               Tony Anthony    actor   
...                           ...   ...                        ...      ...   
3219011            Ever Since Eve  1921                Ethel Lynne  actress   
3405700              Weitertanzen  2008             Rosa Maria Paz  actress   
3427715               The Godsend  1980           Angela Pleasence  actress   
3567516                Motherland  1927            Dorinea Shirley  actress   
3746949               The Seventh  2017  Kerry (II) Willison-Parry  actress   

            character    n  
25954    The Stranger  2.0  
35748    The Stranger  2.0  
46418    The Stranger  NaN  
67421    The Stranger  1.0  
67424    The Stranger  1.0  
...               ...  ...  
3219011  The Stranger  7.0  
3405700  The Stranger  NaN  
3427715  The Stranger  3.0  
3567516  The Stranger  NaN  
3746949  The Stranger  NaN  

[212 rows x 6 columns]

## They are 212  people have played a role called "The Stranger"

### How many roles has Sidney Poitier played throughout his career?

In [30]:
cast[cast['name']== "Sidney Poitier"].describe()

year          n
count    43.000000  42.000000
mean   1965.720930   2.761905
std      11.061141   3.727351
min    1947.000000   1.000000
25%    1957.500000   1.000000
50%    1965.000000   1.500000
75%    1971.000000   3.000000
max    1997.000000  21.000000

## Sidney Poitier played 43 throughout his career

### How many roles has Judi Dench played?

In [31]:
cast[cast['name']== "Judi Dench"].describe()

year          n
count    55.000000  51.000000
mean   1999.672727   8.823529
std      15.120570   9.307429
min    1964.000000   1.000000
25%    1995.000000   2.500000
50%    2004.000000   6.000000
75%    2011.000000  12.000000
max    2018.000000  40.000000

## Judi Dench played 55 roles 

### List the supporting roles (having n=2) played by Cary Grant in the 1940s, in order by year.

In [32]:
ra = cast[cast['character']=='Cary Grant']
ra[ra['year']>=1940]

title  year                name   type   character  \
343376            Broken Promise  2006       Merrill Capps  actor  Cary Grant   
871405   George White's Scandals  1945          Cary Grant  actor  Cary Grant   
871466      Without Reservations  1946          Cary Grant  actor  Cary Grant   
2470956              The Aviator  2004  Michael-John Wolfe  actor  Cary Grant   

          n  
343376  NaN  
871405  NaN  
871466  NaN  
2470956 NaN

## They are no supporting roles (having n=2) played by Cary Grant in the 1940s, in order by year

### List the leading roles that Cary Grant played in the 1940s in order by year.

In [33]:
g = cast[cast['character']=='Cary Grant']
g[(g['year']>1940) & (g['year']<1950)]

title  year        name   type   character   n
871405  George White's Scandals  1945  Cary Grant  actor  Cary Grant NaN
871466     Without Reservations  1946  Cary Grant  actor  Cary Grant NaN

### How many roles were available for actors in the 1950s?

In [34]:
g = cast[cast['type']=='actor']
g[(g['year']>1950) & (g['year']<1960)]

title  year                  name   type  \
66                   Fire Down Below  1957  'Stretch' Cox Troupe  actor   
68                Dorp aan de rivier  1958         Frans 't Hoen  actor   
111             Ci sposeremo a Capri  1956         Quartetto 2+2  actor   
119                   Die Todesarena  1953      Die 3 Motorellos  actor   
289      Sri Kalahastiswara Mahatyam  1954        Subba Rao A.V.  actor   
...                              ...   ...                   ...    ...   
2532460             Salve sauegjeter  1958           ?yvind ?yen  actor   
2532468              La noche avanza  1952       Francisco ?beda  actor   
2532533                    Yaban kiz  1954             Memduh ?n  actor   
2532643      Barbaros Hayrettin Pasa  1951          Ali ?st?ntas  actor   
2532644                     Kara Efe  1952          Ali ?st?ntas  actor   

                   character     n  
66            Limbo Dance by  20.0  
68                  Dirk Jan  15.0  
111               Themselves  19.0  
119               Themselves   NaN  
289              Basavareddy  12.0  
...                      ...   ...  
2532460  Asbj?rn, sauegjeter   2.0  
2532468             Pelotari  21.0  
2532533               Mehmet   4.0  
2532643               Levend   NaN  
2532644             Kara Efe   1.0  

[140978 rows x 6 columns]

## They are 140978 roles were available for actors in the 1950s

### How many roles were available for actresses in the 1950s?

In [35]:
g = cast[cast['type']=='actress']
g[(g['year']>1950) & (g['year']<1960)]

title  year             name     type  \
2532693   De moedige bruidegom  1952   Hilda 't Seyen  actress   
2532730      Cuban Rebel Girls  1959  Beverly Aadland  actress   
2532731    Death of a Salesman  1951  Beverly Aadland  actress   
2532732   Marjorie Morningstar  1958  Beverly Aadland  actress   
2532733          South Pacific  1958  Beverly Aadland  actress   
...                        ...   ...              ...      ...   
3786056         Jeg elsker dig  1957       Clara ?st?  actress   
3786058       Styrmand Karlsen  1958       Clara ?st?  actress   
3786059       Styrmand Karlsen  1958       Clara ?st?  actress   
3786061  Tre piger fra Jylland  1957       Clara ?st?  actress   
3786075                   Toya  1956       Aslaug ?ye  actress   

                          character     n  
2532693       De moeder van Tinneke   NaN  
2532730               Beverly Woods   2.0  
2532731                        Girl   NaN  
2532732                      Dancer   NaN  
2532733  Nurse in Thanksgiving Show  54.0  
...                             ...   ...  
3786056                   Frk. Muhs   9.0  
3786058           Dronning Viktoria   9.0  
3786059                   The Queen   9.0  
3786061                   Frk. Lind   9.0  
3786075            moren til Trygve  19.0  

[52096 rows x 6 columns]

## They are 52096 roles were available for actresses in the 1950s

### How many leading roles (n=1) were available from the beginning of film history through 1980?

In [36]:
g = cast[(cast['year']<1980)]
g.dropna()
g[g['n'] == 1]['n'].value_counts()

1.0    63957
Name: n, dtype: int64

## They are 63957 leading roles (n=1) were available from the beginning of film history through 1980

### How many non-leading roles were available through from the beginning of film history through 1980?

In [37]:
g = cast[(cast['year']<1980)]
g.dropna()
g[g['n'] > 1]['n'].describe()

count    651497.000000
mean         10.043507
std           9.355020
min           2.000000
25%           4.000000
50%           8.000000
75%          13.000000
max         788.000000
Name: n, dtype: float64

## They are 651497 non-leading roles were available through from the beginning of film history through 1980

### How many roles through 1980 were minor enough that they did not warrant a numeric "n" rank?

In [38]:
g = cast[(cast['year']<1980)]
g['n'].isnull().sum()

443881

## They are 443881 roles through 1980 were minor enough that they did not warrant a numeric "n" rank